### ISSUE TITLE
Language Detection overrides language explicitly set in files.associations

### ISSUE THREAD

**Issue by pgit** [microsoft/vscode#133506]
_Does this issue occur when all extensions are disabled?:_
Hard to tell -- both false language detections I observed where from plugins (python and bash), no builtin language yet (see below).

- VS Code Version: 1.60.1
- OS Version: Windows 10 / RHEL8

_Steps to Reproduce:_

1. Its difficult to provide a minimal testcase, as the "guessing" engine of language detection is involved. In each project, we have a file called "Buildfile" which is manually set to language type "makefile" in per-User setting ("files.associations").
2. In some project, the Buildfile is recognized as e.g., "python" or "bash" script.
3. Disabling language detection fixes this.

IMO, in any case, a manually configured lanugage should never be overridden.

**Comment by TylerLeonhardt** [comment]
Can you share a `files.associations` and `Buildfile` that causes the change so I can look into the root cause?

**Comment by pgit** [comment]
Hi, settings contain these file assocations:
```
"files.assocations": {
   "Buildfile": "makefile",
   "Buildfile.global*": "makefile",
    "*.h": "cpp",
    "*.c": "cpp",
    // some more custom file types mapped to "json"
}
````

As to the Buildfile, unfortunatelly I'm not at liberty to share that. It starts with a shebang of some custom build tool and mostly looks like a make file (that's why we chose "makefile" as a language, to get some syntax highlighting).  Some build rules do include bash scripts, that's why it may have been detected as such. But most times, it has been detected as python.

I hope you can make sense of this without the actual file. If there is any debug logging I can enable for this, I'll try to provide. But in the end, I think the actual contents of the it shouldn't matter whatsoever when I'm manually assigning a language to it.

**Comment by bersbersbers** [comment]
Here's a reproducible example. A file named `NAMESPACE` (related to R packages) with the following content keeps getting auto-detected as "Python" despite

```json
    "files.associations": {
        "NAMESPACE": "plaintext"
    },
```

```
import(shiny)
importFrom(ggplot2,aes)
importFrom(ggplot2,coord_cartesian)
importFrom(ggplot2,coord_flip)
importFrom(ggplot2,element_blank)
importFrom(ggplot2,element_rect)
importFrom(ggplot2,element_text)
importFrom(ggplot2,geom_bar)
importFrom(ggplot2,geom_boxplot)
importFrom(ggplot2,geom_jitter)
importFrom(ggplot2,geom_label)
importFrom(ggplot2,geom_line)
importFrom(ggplot2,geom_point)
importFrom(ggplot2,geom_ribbon)
importFrom(ggplot2,geom_text)
importFrom(ggplot2,geom_tile)
```

**Comment by TylerLeonhardt** [comment]
@bersbersbers your example is very interesting for two reasons:

1. we run autodetection on plaintext regardless of if it's set in `files.associations` and it looks like at the auto detect layer I can't tell the difference between a file that has been set to `plaintext` via `files.associations` and one that is just plaintext by default.

2. However, the original issue is that this was happening regardless of what `files.associations` has set (i.e. `makefile`) and it was intermittent and I think I know when this is happnening. The original issue here is actually a race condition when you are loading vscode with the `NAMESPACE` file already opened as the active editor:
  1. add:
```
    "files.associations": {
        "NAMESPACE": "makefile"
    },
```
  2. create a `NAMESPACE` file with the contents @bersbersbers suggested
  3. Notice that it is correctly treated as a makefile
  4. leave the file open and then reload the window
  5. Notice that the `NAMESPACE` file first becomes a makefile and then moments later becomes Python

that's the bug. The bug is that the editor is getting first treated as a plaintext file... which triggers auto-detection... then it's set to makefile as vscode finishes starting up... but auto-detect was already triggered so it gets changed to python.

I'm wondering if I can somehow have language detection wait some amount of time before it's run for the first time... Maybe a global promise that resolves after X seconds that language detection waits on.

@bpasero are you aware of any mechanisms for doing what I wanna do here? Or any advice?

**Comment by bpasero** [comment]
> I'm wondering if I can somehow have language detection wait some amount of time before it's run for the first time... Maybe a global promise that resolves after X seconds that language detection waits on.

We have a way to wait on certain lifecycle phases, via `ILifeCycleService`, but I am not sure I understand the reasoning fully...

**Comment by bersbersbers** [comment]
@TylerLeonhardt you are right that 
```json
    "files.associations": {
        "NAMESPACE": "makefile"
    },
```
works around the issue for me - thanks! Good luck fixing the underlying one.

**Comment by TylerLeonhardt** [comment]
Might be a better candidate for this "wait until ready before running auto-detect for the first time"
```ts
await IExtensionService.whenInstalledExtensionsRegistered
```

**Comment by alexdima** [comment]
:+1: for waiting on `IExtensionService.whenInstalledExtensionsRegistered`. To distinguish between an intentional plaintext and a "fallback" plaintext, I think looking at `ITextMolde.getLanguageIdentifier()` is not enough. I don't think that can distinguish between text buffers intentionally set to plaintext via the status bar or text buffers that use plaintext due to a lack of file name, ext name, or first line match.

**Comment by bpasero** [comment]
I think due to language guessing we now can actually distinguish an explicit set mode:

https://github.com/microsoft/vscode/blob/dbeff882f511cc2a3427ed9370780687668590ce/src/vs/workbench/common/editor/textEditorModel.ts#L83-L84

**Comment by TylerLeonhardt** [comment]
waiting on `whenInstalledExtensionsRegistered` is now checked in to main so the original issue is fixed... however I'm leaving this open for a day to see if I can fix the other issue (plaintext in file association triggering auto-detect)... if I can't fix it in time, I'll make a new issue for the next milestone.

anyway on to that...

@bpasero this is interesting... there seems to be some deeper behavior that I didn't factor in. When you set:
```
"files.associations": {
        "NAMESPACE": "plaintext"
    },
```

I would have expected one of the following:
* The preferredMode is set to `plaintext` on construction: https://github.com/microsoft/vscode/blob/024e05405655ecab3b70938bba4936dcba3de0bc/src/vs/workbench/services/textfile/common/textFileEditorModel.ts#L106
* Something after construction would call [setMode](https://github.com/microsoft/vscode/blob/024e05405655ecab3b70938bba4936dcba3de0bc/src/vs/workbench/services/textfile/common/textFileEditorModel.ts#L197-L201) which would call the `super.setMode` which sets that `hasModeSetExplicitly`

Neither of these cases are true and it seems that [getMode](https://github.com/microsoft/vscode/blob/024e05405655ecab3b70938bba4936dcba3de0bc/src/vs/workbench/services/textfile/common/textFileEditorModel.ts#L978-L984) magically knows (probably looking at `files.associations`) that the file is a X file where X is the value set in `files.associations` (aka `plaintext`... btw, this is also true if I put `makefile` there instead of `plaintext`).

Because of this, it's difficult to differentiate. I _could_ check `files.associations` inside of textFIleEditorModel but it seems hacky.

**Comment by bpasero** [comment]
@TylerLeonhardt that setting is just filling in the "mime" associations (yeah, bad naming, historic reasons) here:

https://github.com/microsoft/vscode/blob/2ab57624591f480742b4230a6d685aa4b5bae530/src/vs/workbench/services/mode/common/workbenchModeService.ts#L176-L183

That registry is being used by the code editor to translate a file path into a mode:

https://github.com/microsoft/vscode/blob/dbeff882f511cc2a3427ed9370780687668590ce/src/vs/workbench/common/editor/textEditorModel.ts#L178

I don't think this should be treated as "preferred mode" or "explicitly set mode" because it is coming from the system and not from the user.

**Comment by TylerLeonhardt** [comment]
> I don't think this should be treated as "preferred mode" or "explicitly set mode" because it is coming from the system and not from the user.

Fair enough. Given that it sounds like I can't really fix this then and it's kinda "by design":

> plaintext in file association triggering auto-detect

**Comment by TylerLeonhardt** [comment]
verification:

1. apply this setting:
```
"files.associations": {
        "NAMESPACE": "makefile"
},
```
2. create a `NAMESPACE` extensionless file with:
```py
importFrom(ggplot2, a)
importFrom(ggplot2, a)
importFrom(ggplot2, a)
importFrom(ggplot2, a)
importFrom(ggplot2, a)
importFrom(ggplot2, a)
importFrom(ggplot2, a)
importFrom(ggplot2, a)
importFrom(ggplot2, a)
importFrom(ggplot2, a)
importFrom(ggplot2, a)
importFrom(ggplot2, a)
importFrom(ggplot2, a)
importFrom(ggplot2, a)
importFrom(ggplot2, a)
importFrom(ggplot2, a)
importFrom(ggplot2, a)
importFrom(ggplot2, a)
importFrom(ggplot2, a)
importFrom(ggplot2, a)
importFrom(ggplot2, a)
importFrom(ggplot2, a)
importFrom(ggplot2, a)
importFrom(ggplot2, a)
importFrom(ggplot2, a)
```
3. now reload the window

At all visible times, the NAMESPACE file should be treated as a `makefile` file and should NEVER be treated as Python.
